### PR TITLE
Fix #2643: Introduce canonical Bazel library structure & migrate util accessibility package

### LINKED ISSUE #2643: Create a BUILD file for accessibility/
This task is part of the build system migration from Gradle to Bazel. A prerequesite before working on this issue is that you are able to build Oppia using Bazel. Check out [Oppia Bazel guide](https://github.com/oppia/oppia-android/wiki/Oppia-Bazel-Guide) for Bazel terminology and more of what it does.

Create a BUILD.bazel file in 

`src/main/java/org/oppia/android/util/accessibility/`

This package does not depend on any other packages inside of Oppia itself. Create a library for each file in the package, if possible, and add the dependency artifacts that are required based on the imports in the file.

## Create the libraries

I'm using the filename interchangebly with the library here.

`AndroidAccessibilityManager.kt` should be a private library since it is only used by `AccessibilityModule.kt`.

`CustomAccessibilityManager.kt` is referenced by `AccessibilityModule.kt`, `FakeAccessibilityModule.kt` and externally. It should be a public library and be a deps dependency of the two first libraries.

## Confirm that the package builds

When you've added the libraries, confirm that you can build the package by using Bazel to build the libraries. If you're in the  directory in your terminal, run

`bazel build :all` 

to build all your libraries.

## Replace srcs dependencies in the module BUILD files

Add the required deps dependencies to the module's BUILD.bazel file.

Since `FakeAccessibilityModule.kt` was also used in some tests, we will have to find the corresponding libraries there as well to replace the srcs dependency with the new deps dependency.

`CustomAccessibilityManager.kt` is used in the app module. Look for the libraries that have `//utility` in its deps dependencies and check if it needs the new library you created as a dep dependency.

## Confirm that the app builds

Now you can confirm that the package works by building the app successfuly.

### PR DESCRIPTION
**Pull request by BenHenning**
Fix #2643.

This is a replacement/continuation of #2668.

This PR does a few things:

1. It introduces a canonical structure for laying out packages with both production & test dependencies (see https://github.com/oppia/oppia-android/pull/2668/files#r577336935) using the accessibility package under the utility module as an example (& as part of fixing #2643).
2. It introduces new top-level package groups in the root BUILD.bazel to simplify visibility management & to provide us with fine tune control (see https://docs.bazel.build/versions/master/be/functions.html#package_group).
3. It refactors the codebase to more consistently switch between prod & test modules for the accessibility checker to ensure dependencies are properly hooked up.
4. It does some other minor cleanups in the new package.
5. It introduces a way to exclude migrated packages (using a glob).

One thing to note: long-term it's expected that modules will only be referenced in test dependencies & the root app binary targets. However, at the moment production modules need to be included in the main ``app`` library due to being tightly coupled with the Dagger component classes. This will continue to be the case until #1720 is finished (since then build graph inclusion of a module predicates whether it's included in the generated component).

### CODE REVIEW

**Comment by BenHenning** [comment]
This replaces #2668.

**Comment by BenHenning** [comment]
NB: I am expecting some Bazel failures, so they will need to be fixed in a follow-up commit once the run finishes.

**Review by rt4914 — APPROVED** [review]
Approving for files for which I am the code-owner. Thanks.

File: utility/BUILD.bazel
Comment (FareesHussain) [inline utility/BUILD.bazel]: We need to exclude the files inside [accessibility](https://github.com/oppia/oppia-android/tree/develop/utility/src/main/java/org/oppia/android/util/accessibility)

I guess you need to add a kdoc here similar to app/BUILD.bazel.

```suggestion
MIGRATED_PROD_FILES = glob(["src/main/java/org/oppia/android/util/accessibility/**/*.kt"])
```

File: utility/src/main/java/org/oppia/android/util/accessibility/AccessibilityCheckerImpl.kt
Comment (FareesHussain) [inline utility/src/main/java/org/oppia/android/util/accessibility/AccessibilityCheckerImpl.kt:17]: I guess you've missed adding the Kdocs for these two functions.

File: BUILD.bazel
Comment (FareesHussain) [inline BUILD.bazel:8]: Isn't this similar to `oppia_binary_visibility` below?

if not what packages does each cover, when do we need to use `oppia_api_visibility` & `oppia_binary_visibility`

File: utility/src/main/java/org/oppia/android/util/accessibility/BUILD.bazel
Comment (FareesHussain) [inline utility/src/main/java/org/oppia/android/util/accessibility/BUILD.bazel]: Do we need to specify this as it is private by default?

**Review by FareesHussain — CHANGES_REQUESTED** [review]
@BenHenning Added some suggestions and comments, PTAL.

File: utility/BUILD.bazel
Comment (BenHenning) [reply utility/BUILD.bazel]: Ah. I'm not sure how I missed this--thanks! Will update it locally a bit later and make sure everything's working as expected.

File: BUILD.bazel
Comment (BenHenning) [reply BUILD.bazel:8]: Hmm I tried making this clear in the documentation, but maybe it could use a bit more work. We'll generally be using API visibility unless something should only be accessible to the binary. I split them out to make the visibility groups themselves more explicit & to provide more control.

File: utility/src/main/java/org/oppia/android/util/accessibility/BUILD.bazel
Comment (BenHenning) [reply utility/src/main/java/org/oppia/android/util/accessibility/BUILD.bazel]: Nope--good point. I'll remove this when I address the other code-based changes.

File: utility/src/main/java/org/oppia/android/util/accessibility/AccessibilityCheckerImpl.kt
Comment (BenHenning) [reply utility/src/main/java/org/oppia/android/util/accessibility/AccessibilityCheckerImpl.kt:17]: I see just one function: ``isScreenReaderEnabled``. Am I maybe missing something?

Also, per https://developer.android.com/kotlin/style-guide#exception_overrides KDocs aren't needed on overrides. In general, we only add them in places where they add value, and method KDocs should always be describing the API introduced by a method. This fits better in the interface or superclass as that's outlining the API implementations need to follow.

File: utility/src/main/java/org/oppia/android/util/accessibility/BUILD.bazel
Comment (BenHenning) [reply utility/src/main/java/org/oppia/android/util/accessibility/BUILD.bazel]: Done.

File: utility/BUILD.bazel
Comment (BenHenning) [reply utility/BUILD.bazel]: Done.

File: utility/src/main/java/org/oppia/android/util/accessibility/AccessibilityCheckerImpl.kt
Comment (FareesHussain) [reply utility/src/main/java/org/oppia/android/util/accessibility/AccessibilityCheckerImpl.kt:17]: Got it

**Review by FareesHussain — APPROVED** [review]
LGTM, thanks.

**Review by anandwana001 — APPROVED** [review]
Approving for files for which I am the code-owner. Thanks.

File: BUILD.bazel
Comment (fsharpasharp) [inline BUILD.bazel:1]: Will we be replacing all instances of //visibility:public with one of these package groups?

**Review by fsharpasharp — APPROVED** [review]

File: utility/src/main/java/org/oppia/android/util/accessibility/BUILD.bazel
Comment (FareesHussain) [inline utility/src/main/java/org/oppia/android/util/accessibility/BUILD.bazel]: Why use `:<file_name>` here 
https://github.com/oppia/oppia-android/blob/72bc8970ad2aa843109ffc5023e71bbc6181f5ea/app/src/main/java/org/oppia/android/app/viewmodel/BUILD.bazel#L21

Here I've used only file name

File: app/BUILD.bazel
Comment (FareesHussain) [inline app/BUILD.bazel:535]: I've missed this earlier.

Why are we adding the deps in app/BUILD.bazel 
and why not utility/BUILD.bazel

Comment (BenHenning) [reply app/BUILD.bazel:535]: Because this is where the dependency is needed. It's not used by anything in utility, and to some extent, that's expected since utility is a collection of utility libraries to be used elsewhere in the codebase.

File: BUILD.bazel
Comment (BenHenning) [reply BUILD.bazel:1]: No, I don't think so. We may want to replace all instances with _a_ package group, but the ones here are specifically corresponding to the types of libraries introduced in this PR.

File: utility/src/main/java/org/oppia/android/util/accessibility/BUILD.bazel
Comment (BenHenning) [reply utility/src/main/java/org/oppia/android/util/accessibility/BUILD.bazel]: Mostly auto-complete. I believe they're equivalent, but it's less code-smell to omit the ':'. Will do that in a follow-up commit.

File: app/BUILD.bazel
Comment (FareesHussain) [reply app/BUILD.bazel:535]: Does that mean if there are any packages, which are directly used in the app module then we can directly add this in the app/BUILD.bazel?

but I think it would be better if we can pass it through utility/BUILD.bazel

Comment (BenHenning) [reply app/BUILD.bazel:535]: Yes, that's how it should be. We're planning to get rid of all the top-level BUILD files--libraries should always depend as directly on the thing they need as possible. Occasionally we may introduce libraries that export parts of their implementation (which can be useful when versioning APIs), but that's something that would come after everything is modularized.

Happy to chat about this further, and I'm sure @fsharpasharp can provide some insight here, too.

File: utility/src/main/java/org/oppia/android/util/accessibility/BUILD.bazel
Comment (BenHenning) [reply utility/src/main/java/org/oppia/android/util/accessibility/BUILD.bazel]: Done.

**Comment by BenHenning** [comment]
I think all of your comments are now addressed @FareesHussain. PTAL & resolve if they are satisfactorily addressed.

**Review by FareesHussain — APPROVED** [review]
LGTM, thanks.

**Comment by BenHenning** [comment]
Thanks! Looking at failures.

**Comment by BenHenning** [comment]
Hmm, Bazel failures seem more of the incorrect affected targets computation. Gonna see if I can repro locally & actually fix the script.

Gradle failure seems like a weird kapt flake I haven't seen before.

**Comment by BenHenning** [comment]
Re-running tests to unblock down-stream work. Would like to try and investigate that Bazel issue, though. It's been eluding me...

**Comment by BenHenning** [comment]
Aha, it appears the Gradle failures are real. Have some fixes to do.

**Pull request merged**
